### PR TITLE
Introduce lsp-completion-passthrough-try-completion

### LINKED ISSUE #1653: company-complete-common not completing common prefix with `company-capf`
```
Before logging the bug, please make sure that:
1. You are using the latest version of `lsp-mode` related packages.
```
Yes, lsp version from melpa `20200504.643`.

```
2. You may also try reproduce the issue using clean environment using the following command:
emacs -q -l lsp-start-plain.el
```
I tried this and the issue is still there. Although one caveat the completions when I use `lsp-start-plain.el` are coming from `company-clang` and I use `company-capf` as recommended by the documentation. So it's might not be representative of the real issue. I have tried `lsp-mode` in Spacemacs however and the problem exists there too.

```
3. Check FAQ and Troubleshooting section (https://github.com/emacs-lsp/lsp-mode#faq and https://github.com/emacs-lsp/lsp-mode#troubleshooting)
```
Nothing in FAQ related to this.

```
Alternatively, in will be great if you can reproduce the issue using [lsp-docker](https://github.com/emacs-lsp/lsp-docker/)  which provides the minimal configurations for `lsp-mode` and ships with most of the language servers.
```
I looked into this but since the docker uses `ccls` for C++ it doesn't apply. I use `clangd` and `company-capf` backend so I thought it won't reproduce my environment so I didn't try it. Still happy to try it but I think it's irrelevant.

```
3. Consider providing a minimal project to reproduce the issue with. 
```
[lsp-company-mode-test.zip](https://github.com/emacs-lsp/lsp-mode/files/4574348/lsp-company-mode-test.zip) is the project, build using `cmake -DCMAKE_EXPORT_COMPILE_COMMANDS=1 .`

```
company-backends '(company-capf)
company-version  0.9.12
completion-at-point-functions == (lsp-completion-at-point)
emacs-version = GNU Emacs 26.3 (build 1, x86_64-apple-darwin18.5.0, Carbon Version 158 AppKit 1671.4) of 2019-08-29)
``` 

**Describe the bug**
Common parts of the completion are not completing with `<tab>` or calling `company-complete-common` directly from the potential candidates. More details https://github.com/company-mode/company-mode/issues/980 the folks over there saying its an `lsp-mode` issue.

More detailed explanation with a gif.

**To Reproduce**
Open the `main.cpp` make sure `clangd` server is used with `company-capf` as the `company-backend`.

Type `Some` at line 15 and press `<tab>`. It doesn't complete to `Something` while it should.
If you use `company-dabbrev-code` as the only `company-backend` you will that it does complete to `Something` properly. So I am assuming something to do with `company-capf` backend.

**Expected behavior**
Type `Some` at line 15 and press `<tab>`. With the recommended `company-capf` backed it should complete to `Something`.

**Which Language Server did you use**
I am using `clangd` with `company-capf` as the only completion backend. `clandg` is called like `clangd -j=8 -background-index -log=verbose`

**OS**
MacOs Catalina 10.15.4

**Error callstack**
No errors produced while trying to complete.

### PR DESCRIPTION
**Pull request by dgutov**
Hi!

The latest version of Company features a tighter integration for `try-completion` behavior of CAPF, and a side-effect is inheriting its problems too. :frowning: See this video:

[Screencast from 2024-08-19 04-57-37.webm](https://github.com/user-attachments/assets/48cdc678-dbb6-4ff4-bc6d-91d99fde69d5)

From [this comment](https://github.com/company-mode/company-mode/pull//1488#issuecomment-2295533862), also see debbugs report linked there.

The same or similar bug is also present when using the default completion UI and Corfu, and this should fix that too.

Note this does not address #1653, but could be the place for the subsequent improvement/workaround.

### CODE REVIEW

File: lsp-completion.el
Comment (kiennq) [inline lsp-completion.el:770]: Since this is a passthrough try completion, can we just return the `(cons string point)` instead of using `try-completion` here?

File: lsp-completion.el
Comment (dgutov) [reply lsp-completion.el:759]: AFAICS lsp-mode doesn't use this variable when matching: the list of completions is constructed case-insensitively, at least with the language server that I'm currently testing with.

File: lsp-completion.el
Comment (dgutov) [reply lsp-completion.el:770]: It's possible - then "expand common" will always just keep the current input.

That seems like a bit of a loss from my POV, but VS Code never supported this anyway, so :shrug:

File: lsp-completion.el
Comment (kiennq) [reply lsp-completion.el:759]: Ah, that's right. `lsp-mode` is using `string-match`, so this should be bound to `case-fold-search` instead

**Comment by kiennq** [comment]
@dgutov can you help me to understand more how `company-mode` is going to use the `try-completion` and how it is going to affect the current completions?

File: lsp-completion.el
Comment (dgutov) [reply lsp-completion.el:759]: Okay. Well, when you do string matching in completion, I think you're expected to bind `case-fold-search` to `completion-ignore-case` around that logic.

Or just go with the value `t`, like I think happens now most of the time.

**Comment by dgutov** [comment]
> @dgutov can you help me to understand more how `company-mode` is going to use the `try-completion` and how it is going to affect the current completions?

It will only be used by `company-complete-common` (the current TAB binding). Where previously that command only did simple prefix matching, now it delegates to try-completion. [Here's the PR](https://github.com/company-mode/company-mode/pull/1488) with an illustration video.

File: lsp-completion.el
Comment (kiennq) [reply lsp-completion.el:770]: Agree, the `basic-try-completion` is also matching prefix anyway, and in case of advanced language server that do the fuzzy match by themselves, I don't think we want to be smart and step in to decide where is the common prefix.

File: lsp-completion.el
Comment (kiennq) [reply lsp-completion.el:759]: Great, I can bind the `case-fold-search` to `completion-ignore-case` in another PR, or if possible, can you help to bind it in this same PR if possible :).
I think there's only two places in this file.

File: lsp-completion.el
Comment (kiennq) [reply lsp-completion.el:770]: This is my main feedback, for other comments, feel free to won't fix them.

**Review by kiennq — APPROVED** [review]

File: lsp-completion.el
Comment (dgutov) [reply lsp-completion.el:759]: Note that `case-fold-search` is `t` by default, whereas `completion-ignore-case` is nil by default.

Obeying the user-set `completion-ignore-case` would make lsp-mode's completion case-sensitive. This is not bad (Eglot currently does that), but it's a significant change in behavior, I think, and it'll make it less "passthrough".

So, are you sure you want that?

Comment (kiennq) [reply lsp-completion.el:759]: I mean we should change that in other places to make it conforms with `completion-ignore-case`. In this passthrough function, just returns `(cons string point)` should be okay. It will be similar to VsCode.
@yyoncho, let me know if you have any concern

Comment (dgutov) [reply lsp-completion.el:759]: > In this passthrough function, just returns (cons string point) should be okay.

Like mentioned, I'd like to keep the change in behavior in this PR to a minimum. After the merge, this code would be equivalent to the latest Eglot in Emacs master.

Except Eglot does case matching for prefix (by default) - and lsp-mode ignores case, not sure which is better but both of the corresponding behaviors seem fit the corresponding package' goals.

Comment (dgutov) [reply lsp-completion.el:759]: You might very well do that change in the next edit, though (if there is agreement among the maintainers).

Comment (kiennq) [reply lsp-completion.el:759]: Got it. I signed off for this change. Let's get another signoff and we can merge

**Pull request merged**
